agent/util-scripts/pbench-tool-meister-stop

```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/env python3
+# -*- mode: python -*-
+
+"""pbench-tool-meister-stop
+
+Responsible for stopping all local/remote tool meisters, closing down the data
+sink, and finally the redis server.
+"""
+
+import sys
+import os
+import json
+import time
+import errno
+import logging
+import redis
+import signal
+
+# port number is "One Tool" in hex 0x17001
+# FIXME: move to common area
+redis_port = 17001
+# FIXME: this should be moved to a shared area
+channel = "tool-meister-chan"
+
+
+def is_running(pid):
+    """Is the given PID running?
+
+    See https://stackoverflow.com/questions/7653178/wait-until-a-certain-process-knowing-the-pid-end
+    """
+    try:
+        os.kill(pid, 0)
+    except OSError as err:
+        if err.errno == errno.ESRCH:
+            return False
+    return True
+
+
+def main(argv):
+    """Main program for the tool meister stop.
+
+    This simply sends the "terminate" message to the redis server so that all
+    connected services, tool-meisters, tool-data-sink, etc. will shutdown.
+    Once all services acknowledge the receipt of the "terminate" message, we
+    declare victory.
+
+    TBD:
+
+    We currently have a mode to "double-check" that the services all stopped
+    by ssh'ing into all the local/remote hosts and inspecting the system for
+    any lingering processes.
+    """
+    PROG = os.path.basename(argv[0])
+    logger = logging.getLogger(PROG)
+    if os.environ.get("_PBENCH_TOOL_MEISTER_STOP_LOG_LEVEL") == "debug":
+        log_level = logging.DEBUG
+    else:
+        log_level = logging.INFO
+    logger.setLevel(log_level)
+    sh = logging.StreamHandler()
+    sh.setLevel(log_level)
+    shf = logging.Formatter("%(message)s")
+    sh.setFormatter(shf)
+    logger.addHandler(sh)
+
+    try:
+        full_hostname = os.environ["full_hostname"]
+    except Exception:
+        logger.exception("failed to fetch parameters from the environment")
+        return 1
+
+    try:
+        redis_server = redis.Redis(host="localhost", port=redis_port, db=0)
+    except Exception as e:
+        logger.error(
+            "Unable to connect to redis server, localhost:%d: %r", redis_port, e
+        )
+        return 2
+
+    try:
+        tm_pids_raw = redis_server.get("tm-pids")
+        if tm_pids_raw is None:
+            logger.error('Tool Meister PIDs key, "tm-pids", does not exist.')
+            return 3
+        tm_pids_str = tm_pids_raw.decode("utf-8")
```

Comment:
why?

Reply:
Redis transfers bytes, the `json` module needs a python3 string object.